### PR TITLE
[release/1.3.1] [CO-354] Add `RequiresNetworkMagic` and modify `ProtocolMagic`

### PR DESCRIPTION
**Pull request by intricate**
## Description

This PR modifies the `ProtocolMagic` data structure and introduces new data types: `RequiresNetworkMagic` and `ProtocolMagicId` (a new type wrapper around an `Int32`). 

`ProtocolMagic` has been modified such that it consists of two fields: `getProtocolMagicId :: ProtocolMagicId` and `getRequiresNetworkMagic :: RequiresNetworkMagic`.

The  `goldenTestJSONDec` and `goldenTestCanonicalJSONDec` tests which were introduced in #3540 have assisted us in ensuring that we've maintained backwards compatibility in our implementation of `ProtocolMagic`.

## Linked issue

[https://iohk.myjetbrains.com/youtrack/issue/CO-354](https://iohk.myjetbrains.com/youtrack/issue/CO-354)
## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: block/bench/Block.hs [hunk 1]
```diff
@@ -24,17 +24,17 @@ import           Pos.Core (Block, BlockHeader, BlockVersionData (..), Body, Body
...
+import           Pos.Crypto (ProtocolMagic (..), ProtocolMagicId (..), RequiresNetworkMagic (..))
 
 import           Test.Pos.Block.Arbitrary.Generate (generateMainBlock)
 
 -- We need 'ProtocolMagic' and 'ProtocolConstants' in order to generate a
 -- 'MainBlock'.
 
 pm :: ProtocolMagic
-pm = ProtocolMagic 0
+pm = ProtocolMagic (ProtocolMagicId 0) NMMustBeNothing
```
Comment (erikd) [inline block/bench/Block.hs]: \o/

**Comment by intricate** [comment]
@ruhatch Oh, for sure!!

File: crypto/Pos/Crypto/Configuration.hs [hunk 2]
```diff
@@ -1,17 +1,128 @@
...
+        "NMMustBeJust"    -> Right NMMustBeJust
+        other   -> Left ("invalid value " <> show other <>
+                         ", acceptable values are NMMustBeNothing | NMMustBeJust")
+
+-- Canonical JSON instances
+instance Monad m => ToJSON m RequiresNetworkMagic where
+    toJSON NMMustBeNothing = pure (JSString "NMMustBeNothing")
+    toJSON NMMustBeJust    = pure (JSString "NMMustBeJust")
+
+instance ReportSchemaErrors m => FromJSON m RequiresNetworkMagic where
```
[the remark this answers is not in the document]
Comment (mhuesch) [reply crypto/Pos/Crypto/Configuration.hs]: Thanks! Was going to force-push and update, but we just got merged 🙃

[the remark this answers is not in the document]
Comment (intricate) [reply crypto/Pos/Crypto/Configuration.hs]: @ruhatch Interesting. Will look to make this change.

**Comment by mhuesch** [comment]
@ruhatch @intricate For better or worse, we began most of this work on `develop` and then cherry-picked it over to `release/1.3.1`. So we've already experienced >=1/2 of the pain of merging 😅 and it shouldn't be too bad to get this over to `develop`. The harder part is convincing ourselves that the code does the same thing 🤔

**Pull request merged**